azure/services/bastionhosts/spec.go

```diff
@@ -40,69 +41,66 @@ type AzureBastionSpec struct {
-	return s.ResourceGroup
-}
-
-// OwnerResourceName is a no-op for bastion hosts.
-func (s *AzureBastionSpec) OwnerResourceName() string {
-	return ""
+// ResourceRef implements azure.ASOResourceSpecGetter.
+func (s *AzureBastionSpec) ResourceRef() *asonetworkv1.BastionHost {
+	return &asonetworkv1.BastionHost{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      s.Name,
+			Namespace: s.Namespace,
+		},
+	}
 }
 
-// Parameters returns the parameters for the bastion host.
-func (s *AzureBastionSpec) Parameters(ctx context.Context, existing interface{}) (parameters interface{}, err error) {
-	if existing != nil {
-		if _, ok := existing.(armnetwork.BastionHost); !ok {
-			return nil, errors.Errorf("%T is not an armnetwork.BastionHost", existing)
-		}
-		// bastion host already exists
-		return nil, nil
+// Parameters implements azure.ASOResourceSpecGetter.
+func (s *AzureBastionSpec) Parameters(ctx context.Context, existingBastionHost *asonetworkv1.BastionHost) (parameters *asonetworkv1.BastionHost, err error) {
+	bastionHost := &asonetworkv1.BastionHost{}
+	bastionHost.Spec = asonetworkv1.BastionHost_Spec{}
+	if existingBastionHost != nil {
+		bastionHost = existingBastionHost
 	}
 
 	bastionHostIPConfigName := fmt.Sprintf("%s-%s", s.Name, "bastionIP")
+	bastionHost.Spec.AzureName = s.Name
+	bastionHost.Spec.Location = ptr.To(s.Location)
+	bastionHost.Spec.Owner = &genruntime.KnownResourceReference{
+		Name: s.ClusterName,
+	}
+	bastionHost.Spec.Tags = infrav1.Build(infrav1.BuildParams{
+		ClusterName: s.ClusterName,
+		Lifecycle:   infrav1.ResourceLifecycleOwned,
+		Name:        ptr.To(s.Name),
+		Role:        ptr.To("Bastion"),
+	})
+	bastionHost.Spec.Sku = &asonetworkv1.Sku{
+		Name: ptr.To(asonetworkv1.Sku_Name(s.Sku)),
+	}
+	bastionHost.Spec.EnableTunneling = ptr.To(s.EnableTunneling)
+	bastionHost.Spec.DnsName = ptr.To(fmt.Sprintf("%s-bastion", strings.ToLower(s.Name)))
 
-	return armnetwork.BastionHost{
-		Name:     ptr.To(s.Name),
-		Location: ptr.To(s.Location),
-		Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
-			ClusterName: s.ClusterName,
-			Lifecycle:   infrav1.ResourceLifecycleOwned,
-			Name:        ptr.To(s.Name),
-			Role:        ptr.To("Bastion"),
-		})),
-		SKU: &armnetwork.SKU{
-			Name: ptr.To(armnetwork.BastionHostSKUName(s.Sku)),
-		},
-		Properties: &armnetwork.BastionHostPropertiesFormat{
-			EnableTunneling: ptr.To(s.EnableTunneling),
-			DNSName:         ptr.To(fmt.Sprintf("%s-bastion", strings.ToLower(s.Name))),
-			IPConfigurations: []*armnetwork.BastionHostIPConfiguration{
-				{
-					Name: ptr.To(bastionHostIPConfigName),
-					Properties: &armnetwork.BastionHostIPConfigurationPropertiesFormat{
-						Subnet: &armnetwork.SubResource{
-							ID: &s.SubnetID,
-						},
-						PublicIPAddress: &armnetwork.SubResource{
-							ID: &s.PublicIPID,
-						},
-						PrivateIPAllocationMethod: ptr.To(armnetwork.IPAllocationMethodDynamic),
+	if s.SubnetID != "" && s.PublicIPID != "" {
```

Comment:
Ah I think this goes back to my confusion about setting empty ARMIDs since the resulting BastionHostIPConfiguration would be invalid, as discussed here: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4143#discussion_r1379362506

Would setting these IPConfigurations with empty ARMIDs cause problems?

Reply:
Does it cause problems now? I'd be surprised if it worked before and doesn't work now with ASO.

Either way, it seems like these will never be empty based on how we create the spec in azure/scope/cluster.go.

Reply:
It didn't cause problems, I just noticed it provided weird outputs in the unit test. I wasn't aware they were never empty though so I don't think it's an issue. I'll revert the change